scripts/generate_black_spots_input.py

```diff
@@ -0,0 +1,374 @@
+import fiona
+from functools import partial
+import itertools
+import logging
+import os
+import pyproj
+import pytz
+import rtree
+from shapely.geometry import mapping, shape, LineString, Point
+from shapely.ops import transform, unary_union
+
+
+logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
+logger = logging.getLogger()
+
+
+def read_roads(roads_shp):
+    """Reads shapefile and extracts roads and projection"""
+    shp_file = fiona.open(roads_shp)
+    return ([shape(road['geometry']) for road in shp_file], shp_file.crs)
+
+
+def get_intersections(roads):
+    """Calculates the intersection points of all roads"""
+    intersections = []
+    for road1, road2 in itertools.combinations(roads, 2):
+        if road1.intersects(road2):
+            intersection = road1.intersection(road2)
+            if 'Point' == intersection.type:
+                intersections.append(intersection)
+            elif 'MultiPoint' == intersection.type:
+                intersections.extend([pt for pt in intersection])
+            elif 'MultiLineString' == intersection.type:
+                multiLine = [line for line in intersection]
+                first_coords = multiLine[0].coords[0]
+                last_coords = multiLine[len(multiLine)-1].coords[1]
+                intersections.append(Point(first_coords[0], first_coords[1]))
+                intersections.append(Point(last_coords[0], last_coords[1]))
+            elif 'GeometryCollection' == intersection.type:
+                intersections.extend(get_intersections(intersection))
+    # The unary_union removes duplicate points
+    return unary_union(intersections)
+
+
+def get_intersection_buffers(roads, intersection_buffer_meters):
+    """Buffers all intersections"""
+    buffered_intersections = [intersection.buffer(intersection_buffer_meters)
+                              for intersection in get_intersections(roads)]
+    # If intersection buffers overlap, union them to treat them as one
+    return unary_union(buffered_intersections)
+
+
+def split_line(line, max_line_meters):
+    """Checks the line's length and splits in half if larger than the configured max"""
+    if line.length <= max_line_meters:
+        return [line]
+
+    half_length = line.length / 2
+    coords = list(line.coords)
+    for idx, point in enumerate(coords):
+        proj_dist = line.project(Point(point))
+        if proj_dist == half_length:
+            return [LineString(coords[:idx + 1]), LineString(coords[idx:])]
+
+        if proj_dist > half_length:
+            mid_point = line.interpolate(half_length)
+            lines = []
+            lines.extend(
+                split_line(LineString(coords[:idx] + [(mid_point.x, mid_point.y)]), max_line_meters)
+            )
+            lines.extend(
+                split_line(LineString([(mid_point.x, mid_point.y)] + coords[idx:]), max_line_meters)
+            )
+            return lines
```

Comment:
I think a slightly shorter alternative could be `return split_line(head_line) + split_line(tail_line)`